### PR TITLE
MNTOR-3157: Preview Deploy Implementation

### PR DESCRIPTION
**Pull request by mansaj**
# References: 
Jira: MNTOR-3157

# Description
The first part of preview deployment:
- Github action that will run on every PR: 
  - generates a database with the PR number as namespace in GCP cloud SQL
  - set up the user, db table `blurts`
- Build and deploy docker image to GCP artifact registry
- Deploy the image to GCP Cloud Run
- Comment the preview URL
- Github action that cleans up all deployed services once this PR closes

What's **not** included or working:
- FxA (login doesn't work), so the current preview deploy should work for landing page / breaches pages

# How to test
- Open a PR
- Close a PR

# Checklist (Definition of Done)
- [x] All acceptance criteria are met.
- [x] Jira ticket has been updated (if needed) to match changes made during the development process.
- [x] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: Dockerfile_with_migrate [hunk 1]
```diff
@@ -0,0 +1,35 @@
+FROM node:20.12-alpine
```
Comment (rhelmer) [inline Dockerfile_with_migrate:1]: Should this have a different name like `Dockerfile.cloudrun` to make it clearer what it's for?

**Review by rhelmer — APPROVED** [review]
lgtm overall, I'm not sure we really need a separate `Dockerfile` (it's mostly duplicated) but I don't think it'll be too hard to keep up to date, we just might want to think about what the name of it will be.

File: .github/workflows/preview_deploy_gcp.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    - name: Use gcloud CLI
+      run: gcloud info
+
+    - name: Generate database on the fly
+      id: db_create
+      # This can fail on the subsequent runs, we can ignore the error (database already exists)
+      continue-on-error: true
+      run: |
+        gcloud sql instances create ${{ env.SERVICE }}-${{ env.TAG }} --tier=db-f1-micro --region=${{ env.REGION }} --database-version=POSTGRES_15 --edition=enterprise
+        gcloud sql users set-password postgres --host=% --instance ${{ env.SERVICE }}-${{ env.TAG }} --password postgres
```
[the remark this answers is not in the document]
Comment (mansaj) [reply .github/workflows/preview_deploy_gcp.yml:36]: Yea the db won't be publicly accessible. Also short lived

File: Dockerfile_with_migrate (hunk 1 above)
Comment (mansaj) [reply Dockerfile_with_migrate:1]: @rhelmer would this be better? I think maybe we can just reuse the one Dockerfile: https://github.com/mozilla/blurts-server/pull/4448/commits/ecc959aeae0a4b3904ca11fd939fdff8f49adfdc

Comment (mansaj) [reply Dockerfile_with_migrate:1]: nvm that didn't work, trying: https://github.com/mozilla/blurts-server/pull/4448/commits/51ed1c1ed2f6878fdcff245e400dbc36167bab5d

Comment (mansaj) [reply Dockerfile_with_migrate]: Hmm, ok so I'm not able to get the same dockerfile to work with conditionals. I think it has to do with the context when `docker build` happens and when `docker run` happens are in different places (here in github, then in cloudrun).. So the only way is to have different `CMD`

Comment (mansaj) [reply Dockerfile_with_migrate]: But I did rename to `Dockerfile.cloudrun`: https://github.com/mozilla/blurts-server/pull/4448/commits/4daec19ed1ad6e777e92d43df1bd53cb8473623e

**Pull request merged**